### PR TITLE
Contributing Guidelines

### PR DESCRIPTION
**Pull request by dlesnoff**
I wrote a simple CONTRIBUTE.md from the Python project and adapted it to Nim.

I advise for the use of snake_case for file_names, while I recommend camelCase and PascalCase for objects to follow [NEP1 specification](https://nim-lang.org/docs/nep1.html) of the language.

Third party modules is an empty file, but I add it in case, we need third party modules in the future.

I removed references to Gitter and check that the Discord link is correct in each file.

I added a MIT License similar to any of TheAlgorithms project with the correct year.

The README.md is minimalistic. If you have any enhancement that I can do, it might be better to take care of this in another PR, just so we show that the repository is already open to contributions.

- [ ] a better document structure
- [x] the use of Option
- [x] a relaxation on variable naming. We need to define use cases.
- [x] Effect system: use of `funcs` and raises pragma
- [x] Use of `unittest` and `runnableExamples`
- [x] Precise the necessity for a header with a correct file name
- [x] Specify that file names should be looked up in other repositories of The Algorithms
- [x] Specify compiler version (stable?, 2.0 rc1? -> Dev should add at least in PR description, the compiler version. In code too?)
- [x] Export marker. Mandatory?
I think it also lacks links to Nim resources like Nim 2.0 documentation.

### CODE REVIEW

File: CONTRIBUTING.md
Comment (pietroppeter) [inline CONTRIBUTING.md:56]: I feel this is too harsh a constraint. when you have multiple indices `i`, `j`, ... that will appear as indices (summed or transformed in some way) it might make code unreadable. I would find more appropriate a recommendation similar to the one in Python repository:
`Single letter variable names are old school so please avoid them unless their life only spans a few lines.`

File: CONTRIBUTING.md
Comment (pietroppeter) [inline CONTRIBUTING.md:39]:
```suggestion
* have intuitive object and function names that make their purpose clear to readers
```

File: CONTRIBUTING.md
Comment (pietroppeter) [inline CONTRIBUTING.md:43]: if we want to make sure that every function accounts for all possible exceptions we can ask to add a `{.push raises: [].}` at the top of every module as adopted in Status style guide: https://status-im.github.io/nim-style-guide/errors.exceptions.html?highlight=raise#exceptions-errorsexceptions. This would force every user to annotate every `func` with `raises` pragma. An example of a the syntax for raise pragma would be useful.

If the procedure is a func and the push raises pragma above is implemented there is no need to track other side effects.

File: CONTRIBUTING.md
Comment (pietroppeter) [inline CONTRIBUTING.md:44]:
```suggestion
* have documentation comments with clear explanations and/or URLs to source materials
```

I think the term docstring is strictly Python and not used in Nim

File: CONTRIBUTING.md
Comment (pietroppeter) [inline CONTRIBUTING.md:34]:
```suggestion
* have minimal side effects (Examples of side effects: `echo()`, `plot()`, `read()`, `write()`).
```

File: README.md
Comment (pietroppeter) [inline README.md:49]: you probably could already add an empty DIRECTORY.md file (or with some placeholder comment). Not really important since it will likely be fixed by a subsequent PR.

File: CONTRIBUTING.md
Comment (pietroppeter) [inline CONTRIBUTING.md:45]:
```suggestion
* contain `runnableExamples` that test both valid and erroneous input values
```

**Review by pietroppeter — CHANGES_REQUESTED** [review]
great work on this, thanks! I added some comments (mostly adjusting some names common in Python but not in nim).

There is only one thing I honestly object to and I would like to see changed, which is to forbid the use of `i` for indices and the strong message about single letter variable names. I honestly think it is a bit too much (especially for indices) and might make the code less readable. My fear is that would either discourage contributors or you will end up having to relax this constraint. You do want constraint that you actually are willing to enforce.

Another comment on the request of no side effects and tracking of exceptions. I think this is an interesting request to make (I like it) and it makes sense given the codebase focus on algorithms. I am interested to see how easy/hard will be to enforce it. I actually would expect most of the algorithms targeted here to be able to use `func`. This will mean though that almost all contributors will likely need to annotate functions with pragmas, which might be a source of frictions for beginners (could be mitigated with good explanations and we could work on those later). I do expect a number of functions to raise ValueError, although I checked that if you use range values like `Natural` a `ValueError` is not required (nice!).

File: CONTRIBUTING.md
Comment (ZoomRmc) [reply CONTRIBUTING.md:45]: This is a very debatable point. Can't find the discussion, but I'd say there's no consensus on whether or not attaching the tests directly to each proc/func is better than having a test set under `when isMainModule`, specifically, in case of short self-contained modules we expect to be written here.

In any case, it needs to be clarified if these guidelines expect per-proc or top-level `runnableExamples` section.

**Comment by ZoomRmc** [comment]
At this point Nim-specific coding style requirements are listed in three separate paragraphs: `Algorithms should`, `Coding Style` and `Other Requirements for Submissions`. I suggest resolving the reviews above and reorganize the paragraphs.

I propose:
* Algorithm implementation requirements  # for general stuff
* Nim coding style # language-specifics
* Submission requirements
* Optional suggestions

File: CONTRIBUTING.md
Comment (dlesnoff) [reply CONTRIBUTING.md:56]: I find very vague the "only spans a few lines" and did not want to review dozens of code with imprecise, bad variable naming.
Actually too long names can make a code more dense and hard to read indeed.

I might list a few use cases:
multiple indices (matrix, sums integrals and other variables from mathematical formula given as a reference) and (c for character if it is the only character in the scope for string algorithms).

File: CONTRIBUTING.md
Comment (dlesnoff) [reply CONTRIBUTING.md:43]: Why would one annotate every `func` with `raises` pragma ? I am not sure to understand this one.
func is already a proc annotated with {.noSideEffect.}. Isn't an exception a side effect ?
Why would one want to additionally add a pragma to funcs, shouldn't we call it a proc in this case ?
I do not trust the style guide of Status, it looks quite old.

Isn't this something that we will want to remove, once that Nim 2.0 is out ?

**Review by ZoomRmc — CHANGES_REQUESTED** [review]

**Review by ZoomRmc — APPROVED** [review]

File: README.md
Comment (dlesnoff) [reply README.md:49]: I link this subsequent PR #5 is going to solve the problem. It proposes to automate the DIRECTORY.md management.

**Comment by dlesnoff** [comment]
I am not done, it is a little bit too soon to merge it. It still lacks:
- a better document structure
- the use of Option
- a relaxation on variable naming. We need to define use cases.
- Effect system: use of `funcs` and raises pragma
- Use of `unittest` and `runnableExamples`
I think it also lacks links to Nim resources like Nim 2.0 documentation.

File: CONTRIBUTING.md
Comment (pietroppeter) [reply CONTRIBUTING.md:44]: I think exceptions are tracked separately from side effects and a func in general can raise side effects. As far as I know status style guide is kept up to date and used. Not saying it should be used as a style guide here, I just happened to learn about that trick there and think is a nice one.

**Comment by pietroppeter** [comment]
I did not find it mentioned in the guidelines but I realized that the site needs to have a way to understand which file solves this algorithm (in order to associate the various implementations of the algorithms together) so I guess names for algorithm are standardized, correct? I assume the Python version should be used as reference for name of algorithm? I guess this should be mentioned somewhere.

**Comment by dlesnoff** [comment]
Should we enforce the use of the export marker `*`?
It enables to distinguish the helper functions from those that really solves the problem and are part of a user API.
Also, I noticed this sentence in the original CONTRIBUTING.md:
"Algorithms should be packaged in a way that would make it easy for readers to put them into larger programs.".

**Comment by dlesnoff** [comment]
Which version of the compiler is better?
Should we simply recommend to compile against the `stable` version?
Should we put the Nim version in source code?

**Comment by ZoomRmc** [comment]
Learning materials should definitely target stable.
Marking code with the compiler version numbers just adds another thing to manage. Ideally, automated testing should make it unnecessary.
Exporting the main function is fine as a suggestion, in my opinion. No harm if it's omitted.

File: CONTRIBUTING.md
Comment (dlesnoff) [reply CONTRIBUTING.md]: I added an example and explained a little bit the usage of this pragma

File: CONTRIBUTING.md
Comment (dlesnoff) [reply CONTRIBUTING.md]: I finally put both on them at this line. This should be fine now.

File: CONTRIBUTING.md
Comment (dlesnoff) [reply CONTRIBUTING.md]: I was very explicit about this use, and relax greatly the use of single letter variable names.
This section is a little bit long, now !

**Comment by dlesnoff** [comment]
I think titles and document structure can still be greatly enhanced. I have put every information in there. I have tried to follow @ZoomRmc recommandation, but it is not easy to separate Nim information from general recommandation. Especially from a non-native English speaker like me.

File: CONTRIBUTING.md
Comment (Panquesito7) [inline CONTRIBUTING.md:3]:
```suggestion
Welcome to [TheAlgorithms/Nim](https://github.com/TheAlgorithms/Nim)! Before you send your pull requests, please make sure that you **read the whole guidelines**. If you have any doubt about the contributing guide, please feel free to [state it clearly in an issue](https://github.com/TheAlgorithms/Nim/issues/new) or ask the community in [Discord](https://the-algorithms.com/discord).
```

File: CONTRIBUTING.md
Comment (Panquesito7) [inline CONTRIBUTING.md:9]:
```suggestion
We are very happy that you are considering implementing algorithms and data structures for others!\
This repository is referenced and used by learners from all over the globe. Being one of our contributors, you agree and confirm that:
```

This should look something similar to the following (`\` stands for a newline).

---

We are very happy that you are considering implementing algorithms and data structures for others!\
This repository is referenced and used by learners from all over the globe. Being one of our contributors, you agree and confirm that:

File: CONTRIBUTING.md
Comment (Panquesito7) [inline CONTRIBUTING.md:16]:
```suggestion
**New implementations** are welcome! For example, new solutions for a problem, different representations for a graph data structure, or algorithm designs with different complexity but **identical implementation** of an existing implementation are not allowed. Please check whether the solution is already implemented or not before submitting your pull request.
```

File: CONTRIBUTING.md
Comment (Panquesito7) [inline CONTRIBUTING.md:18]: `**` is preferred over `__`.
```suggestion
**Improving comments__ and **writing proper tests** are also highly welcome.
```

File: README.md
Comment (Panquesito7) [inline README.md:45]:
```suggestion
We are on [Discord](https://the-algorithms.com/discord) and [Gitter](https://gitter.im/TheAlgorithms)! Community channels are a great way for you to ask questions and get help. Please join us!
```

File: README.md
Comment (Panquesito7) [inline README.md:16]:
```suggestion
  <a href="https://the-algorithms.com/discord">
```

File: README.md
Comment (Panquesito7) [inline README.md:32]: These can be removed as we don't have CI yet, nor do we use pre-commit or the Python formatter.
```suggestion
```

File: CONTRIBUTING.md
Comment (Panquesito7) [inline CONTRIBUTING.md:28]:
```suggestion
An Algorithm is one or more functions (or classes) that:

```

File: CONTRIBUTING.md
Comment (Panquesito7) [inline CONTRIBUTING.md:36]:
```suggestion
Algorithms should:

```

File: CONTRIBUTING.md
Comment (Panquesito7) [inline CONTRIBUTING.md:46]:
```suggestion
Please focus hard on the naming of functions, objects, and variables.  Help your reader by using **descriptive names** that can help you to remove redundant comments.
```

File: CONTRIBUTING.md
Comment (Panquesito7) [inline CONTRIBUTING.md:100]:
```suggestion
#### Submissions Requirements

- The file extension for algorithm files should be `.nim`.
```

File: CONTRIBUTING.md
Comment (Panquesito7) [inline CONTRIBUTING.md:111]:
```suggestion
  - **Be consistent in the use of these guidelines when submitting.**
  - **Join** us on [Discord](https://the-algorithms.com/discord) **now!**
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Looks great! 🚀
What's the `third_party_modules.md` file for? 🤔

File: CONTRIBUTING.md
Comment (dlesnoff) [reply CONTRIBUTING.md:16]: Identical implementation is singular if I am not mistaken, so *is* not allowed looks correct to me.

File: CONTRIBUTING.md
Comment (dlesnoff) [reply CONTRIBUTING.md:9]: I commit this, but I prefer blank spaces than visible characters that are escaped.
People may read the CONTRIBUTING.md in CLI directly rather than in GitHub.

**Comment by dlesnoff** [comment]
> What's the third_party_modules.md file for?

That's for optional packages not in the standard library.

File: CONTRIBUTING.md
Comment (Panquesito7) [inline CONTRIBUTING.md:17]:
```suggestion
**Improving comments** and **writing proper tests** are also highly welcome.
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Review by Panquesito7 — APPROVED** [review]
Looking great. Thank you! 🚀

We can always improve it later on. Will just wait until the other reviewers approve it in case anything was missed.
By the way, some of the badges are not showing up due to some issue with `img.shields` apparently.

**Comment by dlesnoff** [comment]
Well @pietroppeter approved these changes right before those last small markdown fixes.
I would like to get @ZoomRmc review before merging just in case.
It is indeed important to merge it quickly so that the repository has a front page and to attract more contributors.

Specific refinements of CONTRIBUTING.md can be made later.

**Pull request merged**
